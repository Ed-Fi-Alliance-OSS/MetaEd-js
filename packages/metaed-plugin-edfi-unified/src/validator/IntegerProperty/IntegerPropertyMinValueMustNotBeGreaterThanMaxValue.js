// @flow
import type { MetaEdEnvironment } from '../../../../../packages/metaed-core/src/MetaEdEnvironment';
import type { ValidationFailure } from '../../../../../packages/metaed-core/src/validator/ValidationFailure';
import { IntegerProperty, IntegerPropertySourceMap, asIntegerProperty } from '../../../../../packages/metaed-core/src/model/property/IntegerProperty';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.propertyIndex.integer.forEach(integer => {
    const integerProperty: IntegerProperty = asIntegerProperty(integer);
    const minValue: number = Number.parseInt(integerProperty.minValue || '0', 10);
    const maxValue: number = Number.parseInt(integerProperty.maxValue || '0', 10);
    if (minValue <= maxValue) return;

    failures.push({
      validatorName: 'IntegerPropertyMinValueMustNotBeGreaterThanMaxValue',
      category: 'error',
      message: `${integer.type} ${integer.metaEdName} has min value greater than max value.`,
      sourceMap: ((integer.sourceMap: any): IntegerPropertySourceMap).minValue,
      fileMap: null,
    });
  });
  return failures;
}
