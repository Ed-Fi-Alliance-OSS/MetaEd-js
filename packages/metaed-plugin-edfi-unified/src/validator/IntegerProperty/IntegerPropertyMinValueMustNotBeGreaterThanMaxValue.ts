// SPDX-License-Identifier: Apache-2.0
// Licensed to the Ed-Fi Alliance under one or more agreements.
// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
// See the LICENSE and NOTICES files in the project root for more information.

import { IntegerProperty, IntegerPropertySourceMap, MetaEdEnvironment, ValidationFailure } from '@edfi/metaed-core';
import { asIntegerProperty } from '@edfi/metaed-core';

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];
  metaEd.propertyIndex.integer.forEach((integer) => {
    const integerProperty: IntegerProperty = asIntegerProperty(integer);
    const minValue: number = Number.parseInt(integerProperty.minValue || '0', 10);
    const maxValue: number = Number.parseInt(integerProperty.maxValue || '0', 10);
    if (minValue <= maxValue) return;

    failures.push({
      validatorName: 'IntegerPropertyMinValueMustNotBeGreaterThanMaxValue',
      category: 'error',
      message: `${integer.type} ${integer.metaEdName} has min value greater than max value.`,
      sourceMap: (integer.sourceMap as IntegerPropertySourceMap).minValue,
      fileMap: null,
    });
  });
  return failures;
}
