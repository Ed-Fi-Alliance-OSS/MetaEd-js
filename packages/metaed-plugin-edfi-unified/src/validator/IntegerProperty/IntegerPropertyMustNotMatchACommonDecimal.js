// @flow
import type { MetaEdEnvironment, ValidationFailure } from '../../../../../packages/metaed-core/index';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.propertyIndex.integer.forEach(integer => {
    metaEd.entity.sharedDecimal.forEach(sharedDecimal => {
      if (integer.metaEdName !== sharedDecimal.metaEdName) return;
      failures.push({
        validatorName: 'IntegerPropertyMustNotMatchACommonDecimal',
        category: 'error',
        message: `${integer.type} ${integer.metaEdName} has the same name as a Common Decimal.`,
        sourceMap: integer.sourceMap.metaEdName,
        fileMap: null,
      });
    });
  });
  return failures;
}
