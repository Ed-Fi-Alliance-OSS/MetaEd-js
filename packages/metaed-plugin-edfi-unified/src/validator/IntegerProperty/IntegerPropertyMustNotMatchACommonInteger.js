// @flow
import type { MetaEdEnvironment, ValidationFailure } from 'metaed-core';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.propertyIndex.integer.forEach(integer => {
    metaEd.entity.sharedInteger.forEach(sharedInteger => {
      if (integer.metaEdName !== sharedInteger.metaEdName) return;
      failures.push({
        validatorName: 'IntegerPropertyMustNotMatchACommonInteger',
        category: 'error',
        message: `${integer.type} ${integer.metaEdName} has the same name as a Common Integer.`,
        sourceMap: integer.sourceMap.metaEdName,
        fileMap: null,
      });
    });
  });
  return failures;
}
