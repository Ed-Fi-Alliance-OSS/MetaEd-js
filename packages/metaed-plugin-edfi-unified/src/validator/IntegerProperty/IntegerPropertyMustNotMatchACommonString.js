// @flow
import type { MetaEdEnvironment, ValidationFailure } from 'metaed-core';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.propertyIndex.integer.forEach(integer => {
    metaEd.entity.sharedString.forEach(sharedString => {
      if (integer.metaEdName !== sharedString.metaEdName) return;
      failures.push({
        validatorName: 'IntegerPropertyMustNotMatchACommonString',
        category: 'error',
        message: `${integer.type} ${integer.metaEdName} has the same name as a Common String.`,
        sourceMap: integer.sourceMap.metaEdName,
        fileMap: null,
      });
    });
  });
  return failures;
}
