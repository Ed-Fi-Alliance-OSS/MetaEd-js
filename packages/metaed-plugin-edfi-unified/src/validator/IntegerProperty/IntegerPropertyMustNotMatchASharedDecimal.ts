import { MetaEdEnvironment, ValidationFailure, ModelBase } from '@edfi/metaed-core';
import { findFirstEntity } from '@edfi/metaed-core';

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];

  metaEd.propertyIndex.integer.forEach((property) => {
    const referencedEntity: ModelBase | null = findFirstEntity(property.metaEdName, [property.namespace], 'sharedDecimal');

    if (referencedEntity != null) {
      failures.push({
        validatorName: 'IntegerPropertyMustNotMatchASharedDecimal',
        category: 'error',
        message: `${property.type} ${property.metaEdName} has the same name as a Shared Decimal.`,
        sourceMap: property.sourceMap.metaEdName,
        fileMap: null,
      });
    }
  });
  return failures;
}
