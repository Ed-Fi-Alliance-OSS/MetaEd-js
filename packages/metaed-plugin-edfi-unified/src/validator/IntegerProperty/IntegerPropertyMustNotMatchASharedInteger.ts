import { MetaEdEnvironment, ValidationFailure, ModelBase } from 'metaed-core';
import { findFirstEntity } from 'metaed-core';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];

  metaEd.propertyIndex.integer.forEach(property => {
    const referencedEntity: ModelBase | null = findFirstEntity(
      property.metaEdName,
      [property.namespace, ...property.namespace.dependencies],
      'sharedInteger',
    );

    if (referencedEntity != null) {
      failures.push({
        validatorName: 'IntegerPropertyMustNotMatchASharedInteger',
        category: 'error',
        message: `${property.type} ${property.metaEdName} has the same name as a Shared Integer.`,
        sourceMap: property.sourceMap.metaEdName,
        fileMap: null,
      });
    }
  });
  return failures;
}
