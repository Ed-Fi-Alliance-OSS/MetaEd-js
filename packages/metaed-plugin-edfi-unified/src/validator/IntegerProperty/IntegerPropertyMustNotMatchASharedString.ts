import { MetaEdEnvironment, ValidationFailure, ModelBase } from '@edfi/metaed-core';
import { findFirstEntity } from '@edfi/metaed-core';

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];
  metaEd.propertyIndex.integer.forEach((property) => {
    const referencedEntity: ModelBase | null = findFirstEntity(property.metaEdName, [property.namespace], 'sharedString');

    if (referencedEntity != null) {
      failures.push({
        validatorName: 'IntegerPropertyMustNotMatchASharedString',
        category: 'error',
        message: `${property.type} ${property.metaEdName} has the same name as a Shared String.`,
        sourceMap: property.sourceMap.metaEdName,
        fileMap: null,
      });
    }
  });
  return failures;
}
