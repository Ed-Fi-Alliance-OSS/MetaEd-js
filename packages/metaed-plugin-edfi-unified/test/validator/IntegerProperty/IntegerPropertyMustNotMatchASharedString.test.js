// @flow
import {
  newMetaEdEnvironment,
  MetaEdTextBuilder,
  DomainEntityBuilder,
  SharedStringBuilder,
  NamespaceBuilder,
} from 'metaed-core';
import type { MetaEdEnvironment, ValidationFailure } from 'metaed-core';
import { validate } from '../../../src/validator/IntegerProperty/IntegerPropertyMustNotMatchASharedString';

describe('when validating integer property does not match shared string', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  let failures: Array<ValidationFailure>;
  let coreNamespace: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi', 'ProjectExtension')
      .withStartSharedString('EntityName1')
      .withDocumentation('EntityDocumentation')
      .withEndSharedString()

      .withStartAbstractEntity('EntityName2')
      .withDocumentation('EntityDocumentation')
      .withIntegerIdentity('PropertyName', 'PropertyDocumentation', '10', '2')
      .withEndAbstractEntity()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityBuilder(metaEd, []))
      .sendToListener(new SharedStringBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('edfi');
    failures = validate(metaEd);
  });

  it('should build one shared string', () => {
    expect(coreNamespace.entity.sharedString.size).toBe(1);
  });

  it('should build one abstract entity', () => {
    expect(coreNamespace.entity.domainEntity.size).toBe(1);
  });

  it('should have no validation failures', () => {
    expect(failures).toHaveLength(0);
  });
});

describe('when validating integer property matches shared string', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const entityName: string = 'EntityName';
  let failures: Array<ValidationFailure>;
  let coreNamespace: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi', 'ProjectExtension')
      .withStartSharedString(entityName)
      .withDocumentation('EntityDocumentation')
      .withEndSharedString()

      .withStartAbstractEntity('EntityName')
      .withDocumentation('EntityDocumentation')
      .withIntegerIdentity(entityName, 'PropertyDocumentation', '10', '2', '0', '10')
      .withEndAbstractEntity()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityBuilder(metaEd, []))
      .sendToListener(new SharedStringBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('edfi');
    failures = validate(metaEd);
  });

  it('should build one shared string', () => {
    expect(coreNamespace.entity.sharedString.size).toBe(1);
  });

  it('should build one abstract entity', () => {
    expect(coreNamespace.entity.domainEntity.size).toBe(1);
  });

  it('should have no validation failures', () => {
    expect(failures).toHaveLength(1);
    expect(failures[0].validatorName).toBe('IntegerPropertyMustNotMatchASharedString');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot();
    expect(failures[0].sourceMap).toMatchSnapshot();
  });
});
