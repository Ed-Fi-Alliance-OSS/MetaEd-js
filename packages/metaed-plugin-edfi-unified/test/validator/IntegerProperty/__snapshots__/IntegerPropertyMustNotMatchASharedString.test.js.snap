// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when validating integer property matches shared string should have no validation failures 1`] = `"integer EntityName has the same name as a Shared String."`;

exports[`when validating integer property matches shared string should have no validation failures 2`] = `
Object {
  "column": 12,
  "line": 8,
  "tokenText": "EntityName",
}
`;
